Bug/93 update vec to bounded vec

bmacer opened the pull request:
addresses #93.  We may be able to alias some of the generics to make it a bit more readable, but this should pass tests for all of equip, market and core, and should successfully remove #[pallet::without_storage_info]

ilionic commented on pallets/rmrk-equip/src/tests.rs:
```diff
@@ -50,33 +50,53 @@ fn create_base_works() {
-			let fixed_part = FixedPart { id: i, z: 0, src: stb("fixed_part_src") };
-			parts.push(PartType::FixedPart(fixed_part));
-		}
+// no longer needed as we use PartsLimit on BoundedVec
```
good to remove if not needed.

bmacer replied:
[196c9d8](https://github.com/rmrk-team/rmrk-substrate/pull/95/commits/196c9d8165e37bc0e2c93d132f456a0f049ea5f4)

ilionic commented on pallets/rmrk-equip/src/mock.rs:
```diff
@@ -40,21 +40,27 @@ frame_support::construct_runtime!(
 	pub MaxMetadataLength: u32 = 256;
 	pub const MaxRecursions: u32 = 10;
 	pub const ResourceSymbolLimit: u32 = 10;
+	pub const PartsLimit: u32 = 10;
```
only 10 parts? I think should be more allowed. Without proper benchmarking hard to tell but need some extra buffer ( how about 30 - 50 ) ?

bmacer replied:
changing to 50.  didn't really apply any logic to the choice of number (either here or in runtime) just had to pick something.

bmacer replied:
[196c9d8](https://github.com/rmrk-team/rmrk-substrate/pull/95/commits/196c9d8165e37bc0e2c93d132f456a0f049ea5f4)

ilionic reviewed (changes requested):
Great stuff, few minor comments.

HashWarlock commented on pallets/rmrk-core/src/functions.rs:
```diff
@@ -509,11 +525,10 @@ where
+		Children::<T>::remove((parent.0, parent.1), (child.0, child.1));
 	}
 
+	/// I don't think we use this at all? [delete this statement if i'm wrong or this function if not]
```
We don't use it, but it could be used as a helper function for RPC API or something in the future. Either choice should be okay.

bmacer replied:
deleting, can always implement again later if we want to use it.

bmacer replied:
[196c9d8](https://github.com/rmrk-team/rmrk-substrate/pull/95/commits/196c9d8165e37bc0e2c93d132f456a0f049ea5f4)

HashWarlock commented on pallets/rmrk-core/src/functions.rs:
```diff
@@ -568,6 +583,7 @@ where
 		}
 	}
 
+	/// I don't think we use this at all? [delete this line if untrue, or this function if true]
```
Looks like we don't use it. I think it can be removed.

bmacer replied:
removing

bmacer replied:
[196c9d8](https://github.com/rmrk-team/rmrk-substrate/pull/95/commits/196c9d8165e37bc0e2c93d132f456a0f049ea5f4)

HashWarlock commented on pallets/rmrk-core/src/functions.rs:
```diff
@@ -232,11 +249,10 @@ where
 		ensure!(max_recursions > 0, Error::<T>::TooManyRecursions);
```
What happens if we hit the `max_recursions`? Since this is called by `#[transactional]` will that cause the NFT to never get burned from storage since the error will rollback the Storage if this failed?

bmacer replied:
let's say max_recursions is 5, and we have a 10-generation nft being burned.  my belief is that the first 5 generations will successfully burn, and subsequent descendants will fail, but need to test this.  opened #107 to address this.

HashWarlock commented on pallets/rmrk-core/src/functions.rs:
```diff
@@ -232,11 +249,10 @@ where
-		let kids = Children::<T>::take((collection_id, nft_id));
-		for (child_collection_id, child_nft_id) in kids {
-			// Remove child from Children StorageMap
+		let c = Children::<T>::iter_prefix((collection_id, nft_id,)).count();
```
Can we instead  `drain_prefix` and no longer need `remove_child`:
```
for ((child_collection_id, child_nft_id), _) in Children::<T>::drain_prefix((collection_id, nft_id)) {
    Self::nft_burn(child_collection_id, child_nft_id, max_recursions - 1)?;
}
```

ilionic replied:
as @h4x3rotab mentioned in some other PR just `drain_prefix` could be enough in this case, so `Self::nft_burn` not needed?

bmacer replied:
i switch to @HashWarlock's recommendation and it worked.  Will try dropping the Self::nft_burn next.  Would love to avoid recursion if possible.

bmacer replied:
wasn't sure enough about dropping Self::nft_burn just yet.  since the PR is old, we can deal with this separately if needed.

HashWarlock replied:
Sorry I missed this comment. Yeah, the call to `nft_burn` allows for all children & their children to be removed from the `Children` storage. It does involve recursion, but not sure what other strategy since if we remove the `Self::nft_burn` then we will be left will parent-less NFTs left in storage.

HashWarlock reviewed (changes requested):
No longer needed comments can be removed and one request on the `burn_nft` if it is plausible.

bmacer commented on pallets/rmrk-equip/src/lib.rs:
```diff
@@ -41,24 +40,35 @@ pub mod pallet {
 
-		#[pallet::constant]
-		type MaxPartsPerBase: Get<u32>;
+		// No longer needed as we use PartsLimit in BoundedVec now
```
removing

bmacer replied:
[196c9d8](https://github.com/rmrk-team/rmrk-substrate/pull/95/commits/196c9d8165e37bc0e2c93d132f456a0f049ea5f4)

bmacer commented on pallets/rmrk-equip/src/lib.rs:
```diff
@@ -144,7 +153,8 @@ pub mod pallet {
 		AlreadyEquipped,
 		UnknownError,
-		ExceedsMaxPartsPerBase,
+		// no longer needed because we use PartsLimit in BoundedVec
```
removing

bmacer replied:
[196c9d8](https://github.com/rmrk-team/rmrk-substrate/pull/95/commits/196c9d8165e37bc0e2c93d132f456a0f049ea5f4)

HashWarlock commented on pallets/rmrk-core/src/lib.rs:
```diff
@@ -28,11 +28,16 @@ pub type InstanceInfoOf<T> = NftInfo<
 
-pub type ResourceOf<T, R> =
-	ResourceInfo<BoundedVec<u8, R>, BoundedVec<u8, <T as pallet_uniques::Config>::StringLimit>>;
+// pub type ResourceOf<T, R> =
```
We can remove the unused code, right?

bmacer replied:
fab8c5e569c187e450c20120952bf6794bf8d169

HashWarlock commented on pallets/rmrk-core/src/lib.rs:
```diff
@@ -93,8 +107,16 @@ pub mod pallet {
 	/// Stores collections info
-	pub type NftsByOwner<T: Config> =
-		StorageMap<_, Twox64Concat, T::AccountId, Vec<(CollectionId, NftId)>>;
+	pub type NftsByOwner<T: Config> = StorageNMap<
```
Does this duplicate storage with the Uniques pallet's `Owner` StorageNMap? I thought #91 would remove this from storage.

bmacer replied:
going to leave that change for #91, if we're removing it we'll do so there.

HashWarlock commented on pallets/rmrk-core/src/functions.rs:
```diff
@@ -313,9 +319,8 @@ where
-		let kids = Children::<T>::take((collection_id, nft_id));
-		for (child_collection_id, child_nft_id) in kids {
-			// Remove child from Children StorageMap
+		let c = Children::<T>::iter_prefix((collection_id, nft_id,)).count();
```
What is this used for?

bmacer replied:
nothing.  removed. fab8c5e569c187e450c20120952bf6794bf8d169

HashWarlock commented on pallets/rmrk-equip/src/lib.rs:
```diff
@@ -304,12 +308,14 @@ pub mod pallet {
 			let part_length: u32 = parts.len().try_into().unwrap();
-			ensure!(part_length <= T::MaxPartsPerBase::get(), Error::<T>::ExceedsMaxPartsPerBase);
+
+			// We no longer need this because our bound is now on the BoundedVec of parts (PartsLimit)
```
Remove

bmacer replied:
fab8c5e569c187e450c20120952bf6794bf8d169

HashWarlock commented on pallets/rmrk-equip/src/mock.rs:
```diff
@@ -40,21 +40,27 @@ frame_support::construct_runtime!(
 impl pallet_rmrk_equip::Config for Test {
 	type Event = Event;
-	type MaxPartsPerBase = MaxPartsPerBase;
+	// no longer needed as we use PartsLimit on the BoundedVec
```
Remove

bmacer replied:
fab8c5e569c187e450c20120952bf6794bf8d169

HashWarlock commented on runtime/src/lib.rs:
```diff
@@ -336,14 +340,18 @@ parameter_types! {
 	pub const DepositPerByte: Balance = DOLLARS;
 	pub const UniquesStringLimit: u32 = 128;
-	pub const MaxPartsPerBase: u32 = 100;
+	// no longer needed as we use PartsLimit on BoundedVec
```
Remove

bmacer replied:
fab8c5e569c187e450c20120952bf6794bf8d169

HashWarlock commented on runtime/src/lib.rs:
```diff
@@ -336,14 +340,18 @@ parameter_types! {
 impl pallet_rmrk_equip::Config for Runtime {
 	type Event = Event;
-	type MaxPartsPerBase = MaxPartsPerBase;
+	// no longer needed as we use PartsLimit on BoundedVec
```
Remove

bmacer replied:
fab8c5e569c187e450c20120952bf6794bf8d169

HashWarlock commented on traits/src/collection.rs:
```diff
@@ -1,13 +1,16 @@
 /// Collection info.
 #[cfg_attr(feature = "std", derive(PartialEq, Eq))]
-#[derive(Encode, Decode, RuntimeDebug, TypeInfo)]
+// #[derive(Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
```
Remove

bmacer replied:
fab8c5e569c187e450c20120952bf6794bf8d169

HashWarlock reviewed (changes requested):
Remove unused code. I still think the `drain_prefix` should be used for the `burn_nft` function. Maybe @h4x3rotab can clarify if my suggestion is optimal in this situation. Overall looks good though just a few minor changes.

bmacer commented:
> Remove unused code. I still think the `drain_prefix` should be used for the `burn_nft` function. Maybe @h4x3rotab can clarify if my suggestion is optimal in this situation. Overall looks good though just a few minor changes.

@HashWarlock fixed in 390f3c2c3c2c40c8075470265966067f15632791
Couldn't add to burn_nft, because we need to iterate through in order to remove from the Children storage.  but I implemented the drain_prefix inside that loop.  Not sure if that's efficient enough, but that seems required, since we need to iterate through the Children in order to burn, for example, grandchildren.

Cleaned up all the dead code mentioned.

HashWarlock commented:
> > Remove unused code. I still think the `drain_prefix` should be used for the `burn_nft` function. Maybe @h4x3rotab can clarify if my suggestion is optimal in this situation. Overall looks good though just a few minor changes.
> 
> @HashWarlock fixed in [390f3c2](https://github.com/rmrk-team/rmrk-substrate/commit/390f3c2c3c2c40c8075470265966067f15632791) Couldn't add to burn_nft, because we need to iterate through in order to remove from the Children storage. but I implemented the drain_prefix inside that loop. Not sure if that's efficient enough, but that seems required, since we need to iterate through the Children in order to burn, for example, grandchildren.
> 
> Cleaned up all the dead code mentioned.

The `drain_prefix` iterates though so there's no need for the `iter_prefix`. At least this is what I see from the docs https://crates.parity.io/frame_support/storage/types/struct.StorageDoubleMap.html#method.drain_prefix

bmacer commented:
> > > Remove unused code. I still think the `drain_prefix` should be used for the `burn_nft` function. Maybe @h4x3rotab can clarify if my suggestion is optimal in this situation. Overall looks good though just a few minor changes.
> > 
> > 
> > @HashWarlock fixed in [390f3c2](https://github.com/rmrk-team/rmrk-substrate/commit/390f3c2c3c2c40c8075470265966067f15632791) Couldn't add to burn_nft, because we need to iterate through in order to remove from the Children storage. but I implemented the drain_prefix inside that loop. Not sure if that's efficient enough, but that seems required, since we need to iterate through the Children in order to burn, for example, grandchildren.
> > Cleaned up all the dead code mentioned.
> 
> The `drain_prefix` iterates though so there's no need for the `iter_prefix`. At least this is what I see from the docs https://crates.parity.io/frame_support/storage/types/struct.StorageDoubleMap.html#method.drain_prefix

Right, but We need to iterate through the Children to get the Nfts to burn.  We can use `drain_prefix` inside of this (and outside of it), but I don't see how to fully eliminate `iter_prefix` since we're using the data we're iterating through to drain *other* storages.  Maybe I'm missing something?
https://github.com/rmrk-team/rmrk-substrate/blob/ead0e1e8c64ec00518f9472b0facd05c4a010743/pallets/rmrk-core/src/functions.rs#L317-L331

HashWarlock commented:
> > > > Remove unused code. I still think the `drain_prefix` should be used for the `burn_nft` function. Maybe @h4x3rotab can clarify if my suggestion is optimal in this situation. Overall looks good though just a few minor changes.
> > > 
> > > 
> > > @HashWarlock fixed in [390f3c2](https://github.com/rmrk-team/rmrk-substrate/commit/390f3c2c3c2c40c8075470265966067f15632791) Couldn't add to burn_nft, because we need to iterate through in order to remove from the Children storage. but I implemented the drain_prefix inside that loop. Not sure if that's efficient enough, but that seems required, since we need to iterate through the Children in order to burn, for example, grandchildren.
> > > Cleaned up all the dead code mentioned.
> > 
> > 
> > The `drain_prefix` iterates though so there's no need for the `iter_prefix`. At least this is what I see from the docs https://crates.parity.io/frame_support/storage/types/struct.StorageDoubleMap.html#method.drain_prefix
> 
> Right, but We need to iterate through the Children to get the Nfts to burn. We can use `drain_prefix` inside of this (and outside of it), but I don't see how to fully eliminate `iter_prefix` since we're using the data we're iterating through to drain _other_ storages. Maybe I'm missing something?
> 
> https://github.com/rmrk-team/rmrk-substrate/blob/ead0e1e8c64ec00518f9472b0facd05c4a010743/pallets/rmrk-core/src/functions.rs#L317-L331

Maybe I'm confused. I was thinking this would be fine since we would get the child NFTs as we iterate the `Children` using `drain_prefix`. Then we call the `burn_nft` to take care of burning the children and their children as well. Does this cause problems with altering other storages?
```
 for ((child_collection_id, child_nft_id), _) in Children::<T>::drain_prefix((collection_id, nft_id,)) {  
 	Self::nft_burn(child_collection_id, child_nft_id, max_recursions - 1)?; 
 } 
```

HashWarlock commented:
I spoke with @h4x3rotab and the suggestion to replace `iter_prefix` with `drain_prefix` should suffice since we will be recursively calling the `burn_nft` function with the child NFT's collection id and nft id. He also noted that whenever we call `drain_prefix` without utilizing the Iterator that is returned then we can instead call `remove_prefix`. 

Here is an example of this on line 319:
https://github.com/rmrk-team/rmrk-substrate/blob/ead0e1e8c64ec00518f9472b0facd05c4a010743/pallets/rmrk-core/src/functions.rs#L319
Instead we can change it to this and avoid iterating through the Resources:
```
Resources::<T>::remove_prefix((collection_id, nft_id), None);
```

ilionic commented:
Will merge to avoid conflicts, if any adjustments required please create new issue.
